ravedude/src/config.rs

```diff
@@ -0,0 +1,181 @@
+                        NonZeroU32::new(serial_baudrate)
+                            .ok_or_else(|| anyhow::anyhow!("baudrate must not be 0"))?,
+                    ),
+                    None => None,
+                },
+                port: args.port.clone(),
+                reset_delay: args.reset_delay,
+                board: args.board.clone(),
+            },
+            board_config: Default::default(),
+        })
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Debug, Default)]
+#[serde(rename_all = "kebab-case")]
+pub struct RavedudeGeneralConfig {
+    pub open_console: Option<bool>,
+    pub serial_baudrate: Option<NonZeroU32>,
+    pub port: Option<std::path::PathBuf>,
+    pub reset_delay: Option<u64>,
+    pub board: Option<String>,
+}
+
+impl RavedudeGeneralConfig {
+    pub fn apply_overrides(&mut self, args: &crate::Args) -> anyhow::Result<()> {
+        // command line args take priority over Ravedude.toml
+        if args.open_console {
+            self.open_console = Some(true);
+        }
+        if let Some(serial_baudrate) = args.baudrate {
+            self.serial_baudrate = Some(
+                NonZeroU32::new(serial_baudrate)
+                    .ok_or_else(|| anyhow::anyhow!("baudrate must not be 0"))?,
+            );
+        }
+        if let Some(port) = args.port.clone() {
+            self.port = Some(port);
+        }
+        if let Some(reset_delay) = args.reset_delay {
+            self.reset_delay = Some(reset_delay);
+        }
+        Ok(())
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Debug, Default)]
+#[serde(rename_all = "kebab-case")]
+pub struct BoardConfig {
+    pub name: Option<String>,
+    pub inherit: Option<String>,
+    pub reset: Option<ResetOptions>,
+    pub avrdude: Option<BoardAvrdudeOptions>,
+    pub usb_info: Option<BoardUSBInfo>,
+}
+
+impl BoardConfig {
+    pub fn merge(self, base: BoardConfig) -> Self {
+        Self {
+            name: self.name.or(base.name),
+            // inherit is used to decide what BoardGeneralOptions to inherit and isn't used anywhere else
+            inherit: None,
+            reset: self.reset.or(base.reset),
+            avrdude: match self.avrdude {
+                Some(avrdude) => base.avrdude.map(|base_avrdude| avrdude.merge(base_avrdude)),
+                None => base.avrdude,
+            },
+            usb_info: self.usb_info.or(base.usb_info),
+        }
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Debug)]
+pub struct ResetOptions {
+    pub automatic: bool,
+    pub message: Option<String>,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Debug)]
+#[serde(rename_all = "kebab-case")]
+pub struct BoardAvrdudeOptions {
+    pub programmer: Option<String>,
+    pub partno: Option<String>,
+    #[serde(
+        serialize_with = "serialize_baudrate",
+        deserialize_with = "deserialize_baudrate"
+    )]
+    pub baudrate: Option<Option<NonZeroU32>>,
```

Comment:
Can you explain the double `Option` here?

Reply:
The "inner" `Option<NonZeroU32>` signifies the avrdude baudrate, or `None` if it shouldn't be passed (i.e. use the default baudrate). The outer `Option` is there to allow overriding the inherited board's baudrate:
```toml
[board]
inherit = "nano"

[board.avrdude]
# The nano has `baudrate = 57600` but it can be overriden.

# Some(None)
baudrate = -1

# Some(Some(NonZeroU32(8000))
baudrate = 8000

# None
# < no baudrate field set >
```
You can think of it as an `Option<i32>`, but it's an `Option<NonZeroU32>` instead of an `i32` where `-1` is mapped to `None`. So, `Option<Option<NonZeroU32>>`. If this is too convoluted we can change it back into an `i32`.